src/sentry/release_health/metrics.py

```diff
@@ -606,3 +611,131 @@ def extract_row_info(row: Mapping[str, Union[OrganizationId, str]]) -> ReleaseNa
+                crashed = 0
+                metric_id = try_get_string_index(org_id, metric_name)
+                if metric_id is not None:
+                    where = conditions + [
+                        Condition(Column("metric_id"), Op.EQ, metric_id),
+                        Condition(Column("timestamp"), Op.LT, end),
+                    ]
+                    data = raw_snql_query(
+                        Query(
+                            dataset=Dataset.Metrics.value,
+                            match=Entity(entity.value),
+                            select=[Column("value")],
+                            where=where,
+                            groupby=[Column(status_key)],
+                        ),
+                        referrer="release_health.metrics.crash-free-breakdown.session",
+                    )["data"]
+                    for row in data:
+                        if row[status_key] == tag_value(org_id, "init"):
+                            total = int(row["value"])
+                        elif row[status_key] == tag_value(org_id, "crashed"):
+                            crashed = int(row["value"])
+
+                return total, crashed
+
+            sessions_total, sessions_crashed = _get_data(EntityKey.MetricsCounters, "session")
+            users_total, users_crashed = _get_data(EntityKey.MetricsSets, "user")
+
+            return {
+                "date": end,
+                "total_users": users_total,
+                "crash_free_users": 100 - users_crashed / float(users_total) * 100
+                if users_total
+                else None,
+                "total_sessions": sessions_total,
+                "crash_free_sessions": 100 - sessions_crashed / float(sessions_total) * 100
+                if sessions_total
+                else None,
+            }
+
+        return query_stats
+
+    def get_crash_free_breakdown(
+        self,
+        project_id: ProjectId,
+        release: ReleaseName,
+        start: datetime,
+        environments: Optional[Sequence[EnvironmentName]] = None,
+    ) -> Sequence[CrashFreeBreakdown]:
+
+        org_id = self._get_org_id([project_id])
+
+        now = datetime.now(pytz.utc)
+        query_fn = self._get_crash_free_breakdown_fn(
+            org_id, project_id, release, start, environments
+        )
+
+        last: Optional[datetime] = None
+        rv = []
+        for offset in (
+            timedelta(days=1),
+            timedelta(days=2),
+            timedelta(days=7),
+            timedelta(days=14),
+            timedelta(days=30),
+        ):
+            try:
+                item_start = start + offset
+                if item_start > now:
+                    if last is None or (item_start - last).days > 1:
+                        rv.append(query_fn(now))
+                    break
+                rv.append(query_fn(item_start))
+                last = item_start
+            except QueryOutsideRetentionError:
```

Comment:
instead of having dummy_fn, what about letting the tag indexer exceptions bubble up until here?

Reply:
Okay, makes sense. I think we may want an in-process cache around indexer anyway, because we have some endpoints that just call one method on the releasehealth service after the other.

Reply:
Yes, the in-memory cache should be part of the indexer service IMO. I'm writing all my code under the assumption that lookups will be optimized there.